patches/api/0387-Paper-Plugins.patch

```diff
@@ -0,0 +1,1826 @@
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.List;
++
++/**
++ * This class acts as an abstraction for a plugin configuration.
++ */
++public interface PluginConfiguration {
++
++    /**
++     * Gives the name of the plugin. This name is a unique identifier for
++     * plugins.
++     * <ul>
++     * <li>Will only contain alphanumeric characters, underscores, hyphon,
++     *     and period (a-z,A-Z,0-9,_.-).
++     * <li>Typically used for identifying the plugin data folder.
++     * <li>The is the token referenced in {@link #getHardDependencies()}, {@link
++     *     #getSoftDependencies()}, and {@link #getLoadBefore()}.
++     * </ul>
++     * <p>
++     * In the plugin.yml, this entry is named <code>name</code>.
++     * <p>
++     * Example:<blockquote><pre>name: MyPlugin</pre></blockquote>
++     *
++     * @return the name of the plugin
++     */
++    @NotNull
++    String getIdentifier();
++
++    /**
++     * Returns the name of a plugin, including the version.
++     *
++     * @return a descriptive name of the plugin and respective version
++     */
++    @NotNull
++    default String getDisplayName() {
++        return this.getIdentifier() + " v" + this.getVersion();
++    }
++
++    /**
++     * Gives the fully qualified name of the main class for a plugin.
++     * A {@link JavaPlugin} is expected at this location.
++     *
++     * @return the fully qualified main class for the plugin
++     */
++    @NotNull
++    String getMain();
++
++    /**
++     * Gives the phase of server startup that the plugin should be loaded.
++     *
++     * @return phase
++     */
++    @NotNull
++    PluginLoadOrder getLoadPhase();
++
++    /**
++     * Gets the version of this plugin
++     *
++     * @return version string
++     */
++    @NotNull
++    String getVersion();
++
++    /**
++     * Gets the prefix that should be used for the plugin logger.
++     *
++     * @return logger prefix
++     */
++    @NotNull
++    String getLoggingPrefix();
++
++    /**
++     * Gets a list of dependencies that are required for this plugin to load.
++     *
++     * @return immutable list of required dependencies
++     */
++    @NotNull
++    List<String> getHardDependencies();
++
++    /**
++     * Gets a list of dependencies that are used but not required for this plugin to load.
++     *
++     * @return immutable list of soft dependencies
++     */
++    @NotNull
++    List<String> getSoftDependencies();
++
++    /**
++     * Gets a list of dependencies that should be loaded before this plugin is loaded.
++     *
++     * @return immutable list of dependencies to load before
++     */
++    @NotNull
++    List<String> getLoadBefore();
++
++    /**
++     * Gets a list of plugins/dependencies that this plugin provides.
++     *
++     * @return immutable list of provided plugins/dependencies
++     */
++    @NotNull
++    List<String> getProvides();
```

Comment:
getProvidedPlugins()? getAliases()?

Reply:
hmm, aliases might be wrong right ? Technically it may ship some other plugin or smth.
`getProvidedPlugins()` would probs be best ? Tho obviously, someone call the naming experts.